File: ms-productos/src/main/java/com/ferreteria/productos/service/ProductoService.java
package com.ferreteria.productos.service;

import com.ferreteria.productos.dto.ProductoRequest;
import com.ferreteria.productos.dto.ProductoResponse;
import java.util.List;

public interface ProductoService {
    List<ProductoResponse> listarProductos();
    ProductoResponse obtenerPorId(Long id);
    ProductoResponse crear(ProductoRequest request);
    ProductoResponse actualizar(Long id, ProductoRequest request);
    void eliminar(Long id);
    List<ProductoResponse> buscarPorNombre(String nombre);
    List<String> obtenerCategorias();
    List<ProductoResponse> obtenerProductosStockBajo(Integer umbral);
    void descontarStock(Long id, Integer cantidad);
    void restaurarStock(Long id, Integer cantidad);
}

File: frontend-angular/src/app/services/producto.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Producto } from '../models/producto.model';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ProductoService {
  private readonly apiUrl = `${environment.apiUrl}/productos`;

  constructor(private http: HttpClient) {}

  listarProductos(): Observable<Producto[]> {
    return this.http.get<Producto[]>(this.apiUrl);
  }

  obtenerPorId(id: number): Observable<Producto> {
    return this.http.get<Producto>(`${this.apiUrl}/${id}`);
  }

  crear(producto: Omit<Producto, 'idProducto'>): Observable<Producto> {
    return this.http.post<Producto>(this.apiUrl, producto);
  }

  actualizar(producto: Producto): Observable<Producto> {
    return this.http.put<Producto>(`${this.apiUrl}/${producto.idProducto}`, producto);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  actualizarStock(idProducto: number, cantidad: number): Observable<Producto> {
    return this.obtenerPorId(idProducto);
  }

  buscarPorNombre(nombre: string): Observable<Producto[]> {
    return this.http.get<Producto[]>(`${this.apiUrl}/buscar`, { params: { nombre } });
  }

  obtenerCategorias(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/categorias`);
  }

  obtenerProductosStockBajo(umbral: number = 5): Observable<Producto[]> {
    return this.http.get<Producto[]>(`${this.apiUrl}/stock-bajo`, { params: { umbral } });
  }
}
